meta-refkit/classes/image-installer.bbclass

```diff
@@ -0,0 +1,370 @@
+}
+
+python install_images () {
+    import os
+    import shutil
+    import subprocess
+    import stat
+
+    deploy_dir_image = d.getVar('DEPLOY_DIR_IMAGE')
+    install_dir = oe.path.join(d.getVar('IMAGE_ROOTFS'), d.getVar('INSTALLER_IMAGE_DATADIR'))
+    machine = d.getVar('MACHINE')
+    for image, suffix in [x.split(':') for x in d.getVar('INSTALLER_SOURCE_IMAGES').split()]:
+        imagename = '%s-%s.%s' % (image, machine, suffix)
+        path = os.path.join(deploy_dir_image, imagename)
+        if not os.path.exists(path):
+            bb.fatal('Image %s for INSTALLER_SOURCE_IMAGES entry %s:%s not found.' % (path, image, suffix))
+        subprocess.check_output(['install', '-d', install_dir])
+        # Preserve the symbolic links, because that makes images available
+        # under a constant name while also keeping the information about the
+        # exact version of the image which got included.
+        dest = os.path.join(install_dir, imagename)
+        if os.path.islink(path):
+            link_dest = os.path.basename(os.readlink(path))
+            os.symlink(link_dest, dest)
+            os.lchown(dest, 0, 0)
+            dest = os.path.join(os.path.dirname(dest), link_dest)
+            path = os.path.realpath(path)
+        # Hard-linking both saves space during the build *and*
+        # sparseness of the image file. The fallback doesn't, but shouldn't
+        # be needed most of the time.
+        try:
+            os.link(path, dest)
+        except IOError as ex:
+            bb.warn('Hard-linking from %s to %s failed. Falling back to full copy, which looses sparseness: %s' %
+                    (path, dest, str(ex)))
+            shutil.copy2(path, dest)
+        # We run under pseudo, so chown/chmod does not really change the attributes
+        # of the original image file.
+        os.chown(dest, 0, 0)
+        mode = stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH
+        os.chmod(dest, mode)
+}
+do_rootfs[depends] += "${@ ' '.join([x.split(':')[0] + ':do_image_complete' for x in '${INSTALLER_SOURCE_IMAGES}'.split()]) }"
+do_rootfs[postfuncs] += " install_images "
+
+
+# The default installer can be used interactively or
+# configured to run automatically by setting environment
+# variables in advance.
+#
+# The flow is:
+# - determine which image is to be installed
+# - determine which disks are suitable as target, automatically
+#   filtering out the boot disk and everything that is read-only
+#   or completely unusable (removable media)
+# - determine which target to install to
+# - execute installation
+#
+# See below for the commands used by each step.
+#
+# The default installer code is free of bashisms and passed
+# checks with checkbashisms.pl and checkshell. The "local"
+# keyword (although supported by bash and dash) is avoided
+# in favor of wrapping function bodies with local variables
```

Comment:
This says `local` is avoided, but I still see several variables with `local` keyword...